File: test/_zap/run_spider_scan.js
const axios = require('axios').default;

const ZAP_HOST = process.env.ZAP_HOST || 'localhost';
const ZAP_PORT = process.env.ZAP_PORT || 8090;
const CONTEXT_NAME = process.env.CONTEXT_NAME || 'test';


const runSpiderScan = async () => {
  const formData = {
    contextName: CONTEXT_NAME,
    maxChildren: 0,
    recurse: true,
    subtreeOnly: true
  };

  const scanStart = await axios({url: `http://${ZAP_HOST}:${ZAP_PORT}/JSON/spider/action/scan`, method: 'GET',
    params: formData}).catch(err => console.log(err));

  console.log(scanStart.data);
  const scanID = scanStart.data.scan;

  console.log(scanID);

  let percentageComplete = 0;
  let resp = undefined;

  while(percentageComplete < 100) {
    resp = await axios({url: `http://${ZAP_HOST}:${ZAP_PORT}/JSON/spider/view/status`, method: 'GET',
      params: formData});
    console.log(`Spider scan percentage complete: ${percentageComplete}%`);
    percentageComplete = resp.data.status;
  }

  // show final 100% completion
  console.log(`Spider scan percentage complete: ${percentageComplete}%`);

  const scanResults = await axios({url: `http://${ZAP_HOST}:${ZAP_PORT}/JSON/spider/view/fullResults/`,
    method: 'GET',
    params: {
      scanId: scanID
    }}).catch(err => console.log(err));

  scanResults.data.fullResults[0].urlsInScope.forEach(element => {
    console.log('url: ' + element.url + ' status code: ' + element.statusCode);
  });
};

runSpiderScan().then(() => console.log('done')).catch(err => console.log(err));

File: test/_zap/run_spider_scan.py
import os
import requests

ZAP_HOST = os.environ.get('ZAP_HOST') or 'localhost'
ZAP_PORT = os.environ.get('ZAP_PORT') or 8090
CONTEXT_NAME = os.environ.get('CONTEXT_NAME') or 'test'

header = {
    'Accept': 'application/json',
    'X-ZAP-API-Key': None
}

try:
    formData = {
        'contextName': CONTEXT_NAME,
        'maxChildren': 0,
        'recurse': True,
        'subtreeOnly': True
    }

    scanStart = requests.get('http://' + ZAP_HOST + ':' + str(ZAP_PORT) + '/JSON/spider/action/scan',
                             params=formData, headers=header)

    scanID = scanStart.json().get('scan')

    print(scanID)

    percentageComplete = 0
    resp = None

    while(int(percentageComplete) < 100):
        resp = requests.get('http://' + ZAP_HOST + ':' + str(ZAP_PORT) + '/JSON/spider/view/status/?', params={
            'scanId': scanID
        },
            headers=header)

        print('Spider scan percentage complete: ' +
              str(percentageComplete) + '%')
        percentageComplete = resp.json().get('status')

        # show final 100% completion
    print('Spider scan percentage complete: ' + percentageComplete + '%')

    scanResults = requests.get('http://' + ZAP_HOST + ':' + str(ZAP_PORT) + '/JSON/spider/view/fullResults/',
                               params={
                                   'scanId': scanID
    }, headers=header)

    for scannedUrl in scanResults.json().get('fullResults')[0].get('urlsInScope'):
        print('url:' + scannedUrl.get('url') + ' |  status code: ' +
              str(scannedUrl.get('statusCode')))

except requests.exceptions.RequestException as e:
    print(e)
